### DISCUSSION TITLE
Google Summer of Code 2024 Ideas

### DISCUSSION THREAD

**Discussion by louwers** [maplibre/maplibre#318] [General]
We are looking for MapLIbre-related projects & mentors for [Google Summer of Code](https://en.wikipedia.org/wiki/Google_Summer_of_Code) 2024.

Sign-up will begin somewhere in January 2024.

**Comment by louwers** [comment]
@bdon Would you be interested in (co-)mentoring someone to implement PMTiles in C++ and integrate it with MapLibre Native. And as an extended goal to integrate it with MapLibre Native? One downside to this idea: I think there is a good chance someone will do it sooner.

**Reply by bdon** [reply]
@louwers the original PMTiles implementation of [utility functions is C++](https://github.com/protomaps/PMTiles/tree/main/cpp); it's used more extensively than any other implementation right now because it's part of Tippecanoe 2. 

However:

* I think the right approach for Native is to implement an equivalent to JS `addProtocol`. This can be done by exposing interfaces for tiled sources, and a developer can add PMTiles or any other source like Core Data,  or [Contours](https://github.com/onthegomap/maplibre-contour). This is much more powerful, flexible, and vendor-agnostic than supporting PMTiles directly in the code.
* The C++ code does not implement a cache of tile entry directories. This is essential for interactive maps where you fetch one tile at a time. That may be an interesting GSoC project, but...
* I believe (without deep knowledge) that the extensible tiled source interfaces should be exposed in the platform language - Swift or Java, because the behavior for doing HTTP ranges depends on the networking library, and that's platform specific - had a brief chat with @sjg-wdw about this. Disadvantage is anything like Contours or PMTiles would need Java and Swift implementations instead of one C++ one. 

I can't commit to mentoring right now because I don't know the state of contracts, etc. for early next year, but keep me updated!

**Reply by louwers** [reply]
Right! Thanks for sharing. I completely overlooked the C++ utility functions.

> the behavior for doing HTTP ranges depends on the networking library

What do you mean with this exactly? [HttpFileSource](https://github.com/maplibre/maplibre-native/blob/main/src/mbgl/storage/http_file_source.hpp#L11) is the platform aspecific interface which every platform implements, it requests a [Resource](https://github.com/maplibre/maplibre-native/blob/main/include/mbgl/storage/resource.hpp#L13).

This is probably more complicated than I initially thought, especially to do it 'right'. But there's likely an interesting project here. Will get back to you at the start of next year. 👍

**Reply by bdon** [reply]
@louwers yes, I think you're right - I misunderstood the relationship between the C++ bits and the platform. Anyways it's good news if such datasource implementations can be in C++ to avoid needing multiple implementations.

**Reply by louwers** [reply]
@bdon Not sure if you expect to have enough time to be a mentor, but I created an issue for GSoc here: https://github.com/maplibre/maplibre/issues/355

**Comment by sjg-wdw** [comment]
I'm putting together a few proposals for work that include an opening up of the data loading functionality.  If that's accepted, then an implementation for PMTiles would not be very hard for anyone to do.

**Reply by louwers** [reply]
Hi Steve, if this is still an option, I created an issue for GSoC here:

https://github.com/maplibre/maplibre/issues/355
